src/pdb/mod.rs

```diff
@@ -313,37 +300,41 @@ impl Page {
     /// Parse the row groups at the end of the page.
     fn parse_row_groups<R: Read + Seek>(
         reader: &mut R,
-        endian: Endian,
-        args: (PageType, u64, u16, u16, PageFlags),
+        _: Endian,
+        args: (PageType, u64, u16, PageFlags),
     ) -> BinResult<Vec<RowGroup>> {
-        let (page_type, page_heap_offset, num_rows, num_row_groups, page_flags) = args;
-        if num_row_groups == 0 || !page_flags.page_has_data() {
-            return Ok(vec![]);
-        }
+        let endian = Endian::Little;
 
-        let mut row_groups = Vec::with_capacity(num_row_groups.into());
-
-        // Calculate number of rows in last row group
-        let mut num_rows_in_last_row_group = num_rows % RowGroup::MAX_ROW_COUNT;
-        if num_rows_in_last_row_group == 0 {
-            num_rows_in_last_row_group = RowGroup::MAX_ROW_COUNT;
+        let (page_type, page_heap_offset, num_rows, page_flags) = args;
+        if num_rows == 0 || !page_flags.page_has_data() {
+            return Ok(vec![]);
         }
 
-        // Read last row group
-        let row_group = RowGroup::read_options(
-            reader,
-            endian,
-            (page_type, page_heap_offset, num_rows_in_last_row_group),
-        )?;
-        row_groups.push(row_group);
-
-        // Read remaining row groups
-        for _ in 1..num_row_groups {
-            let row_group = RowGroup::read_options(
-                reader,
-                endian,
-                (page_type, page_heap_offset, RowGroup::MAX_ROW_COUNT),
+        let stream_position = reader.stream_position()?;
+
+        // Read row groups
+        let estimated_number_of_row_groups = usize::from(num_rows)
+            .checked_sub(1)
+            .map(|x| x / RowGroup::MAX_ROW_COUNT)
+            .and_then(|x| x.checked_add(1))
+            .ok_or_else(|| binrw::Error::AssertFail {
+                pos: stream_position,
+                message: "Failed to calculate number of row groups".to_string(),
+            })?;
+        let mut row_groups = Vec::with_capacity(estimated_number_of_row_groups);
+        for i in 0..estimated_number_of_row_groups {
+            reader.seek(
+                u64::try_from(row_groups.len())
+                    .ok()
+                    .and_then(|index| index.checked_mul(36))
+                    .and_then(|x| stream_position.checked_sub(x))
+                    .map(SeekFrom::Start)
+                    .ok_or_else(|| binrw::Error::AssertFail {
+                        pos: stream_position,
+                        message: format!("Failed to calculate seek position for row group {}", i),
+                    })?,
```

Comment:
Maybe I'm a little narrow minded from too much C++ programming but this feels overcautious. `row_groups.len()` is a `usize` and will pretty much always fit into a `u64`, right? Are there architectures where usize is bigger than a u64? can we `static_assert` that (is there such a thing in `rust`)?